lib/universe.ts

```diff
@@ -0,0 +1,137 @@
+import path from 'path';
+import { promises as fs } from 'fs';
+import type { Stats } from 'fs';
+
+import Contract from './contract';
+import { UNIVERSE } from './types/types';
+
+/**
+ * @summary Run the callback function concurrently on elements from the given iterator
+ * @function
+ * @memberof module:universe
+ *
+ * Gets up to `concurrency` elements from the given iterator and apply the asynchronous function
+ * concurrently using `Promise.all`.
+ *
+ * If at any point the call to the callback fails, the function will throw the error
+ *
+ * @param it - iterator of elements to go traverse
+ * @param callbackFn - a function to execute for each element in the iterator
+ * @param concurrency - number of elements to apply the function at the same time. Default to 1
+ */
+export async function concurrentForEach<T>(
+	it: IterableIterator<T>,
+	callbackFn: (t: T) => PromiseLike<void>,
+	concurrency = 1,
+): Promise<void> {
+	const run = async () => {
+		const next = it.next();
+		if (next.value && !next.done) {
+			await callbackFn(next.value);
+			await run();
+		}
+	};
+	const runs = [];
+	for (let i = 0; i < concurrency; i++) {
+		runs.push(run());
+	}
+	await Promise.all(runs);
+}
+
+/**
+ * @summary recursively find all files under the directory that match the given filter
+ * @function
+ * @memberof module:universe
+ *
+ * @param dir - base directory to start the search
+ * @param filter - filtering function to indicate that a file should be selected
+ */
+async function findFiles(
+	dir: string,
+	filter: (filePath: string, stat: Stats) => boolean = () => true,
+): Promise<string[]> {
+	const dirFiles = await fs.readdir(dir);
```

Comment:
If you use
```suggestion
	const dirFiles = await fs.readdir(dir, { recursive: true });
```
then you can skip the manual recursion of this function

Reply:
You are right, force of habit. Making the change